keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -514,310 +481,126 @@ spec:
-      quantity: 32Gi
-    - name: productName:
-      string: ACME A4-PCIE-40GB
+      quantity: 16Gi
+    - name: powerSavingSupported
+      bool: true
+    - name: dra.k8s.io/pciRoot # a fictional standardized attribute, not actually part of this KEP
+      string: pci-root-0
 ```
 
-Where "gpu-0" represents one type of card and "gpu-1" represents another (with
-the attributes hanging off each serving to "define" their individual
-properties).
-
-Compared to labels, attributes in this model have values of exactly one type. As
+Compared to labels, attributes have values of exactly one type. As
 described later on, these attributes can be used in CEL expressions to select a
 specific resource for allocation on a node.
 
-While this model is still hypothetical, we do imagine real-world models
-attaching attributes to their resources in a similar way. To avoid any future
-conflicts, we plan to reserve any attributes with the ".k8s.io" suffix for
-future use and standardization by Kubernetes. This could be used to describe
+To avoid any future conflicts, we reserve any attributes with the ".k8s.io" domain suffix
+for future use and standardization by Kubernetes. This could be used to describe
 topology across resources from different vendors, for example, but this is out-
 of-scope for now.
 
-**Note:** If a driver needs to reduce resource capacity, then there is a risk
-that a claim gets allocated using that capacity while the kubelet is updating a
-`ResourceSlice`. The implementations of structured models must handle
-scenarios where more resources are allocated than available. The kubelet plugin
-of a DRA driver ***must*** double-check that the allocated resources are still
+**Note:** If a driver needs to remove a device or change its attributes,
+then there is a risk that a claim gets allocated based on the old
+`ResourceSlice`. The scheduler must handle
+scenarios where more devices are allocated than available. The kubelet plugin
+of a DRA driver ***must*** double-check that the allocated devices are still
 available when NodePrepareResource is called. If not, the pod cannot start until
-the resource comes back. Treating this as a fatal error during pod admission
+the device comes back. Treating this as a fatal error during pod admission
 would allow us to delete the pod and trying again with a new one.
 
 ### Using structured parameters
 
-The following is an example CRD which the developer of the
-`cards.dra.example.com` DRA driver might define as a valid claim parameters
-object for requesting access to its GPUs:
+A ResourceClaim is a request to allocate one or more devices. Each request in a
+claim may reference a pre-defined DeviceClass to narrow down which devices are
+desired or describe that directly in the request itself through a device
+selector. A device selector is a CEL expression that must evaluate to true if a
+device satisfies the request. A special `device` variable provides access to
+the attributes of a device.
```

Comment:
Is there just a single selector per request? At one point we had a design for impliciltly "anding" multiple deviceSelector strings. Is that now not part of the design?

Reply:
A request has multiple requirements, and each requirement is a CEL selector. So you can AND them together.

I need to change "a device selector" into "one or more device selectors".

Reply:
This is related to:
https://github.com/kubernetes/enhancements/pull/4709/files#r1633087176

Reply:
I like that. Changed, with "expression" instead of "deviceSelector".

Reply:
The text still talks about "device selector" because that is what that expression is.

Reply:
What other fields to we expect to live along side `expression`as a one-of in the future?

Reply:
Variable definitions for the CEL expressions was mentioned in one thread somewhere - found it: https://docs.google.com/document/d/1i0xVx884vsWzVaVb72LtEqjsVILmNkeZimAyz4Bj-m0/edit?disco=AAABLY07Bjc